pylint/checkers/unicode.py

```diff
@@ -0,0 +1,713 @@
+                "for some possible problems when using UTF-16 for instance."
+            ),
+        ),
+        "E2402": (
+            "bidirectional unicode can be used to hide real code execution",
+            "bidirectional-unicode",
+            "bidirectional unicode can be used to hide real code execution.",
+        ),
+        "C2403": (
+            "PEP8 recommends UTF-8 as encoding for Python files",
+            "bad-file-encoding",
+            (
+                "PEP8 recommends UTF-8 default encoding for Python files. See "
+                "https://www.python.org/dev/peps/pep-0008/#id20"
+            ),
+        ),
+        **{
+            bad_char.code: (
+                bad_char.description(),
+                bad_char.human_code(),
+                bad_char.help_text,
+            )
+            for bad_char in BAD_CHARS
+        },
+    }
+
+    @staticmethod
+    def _is_invalid_codec(codec: str) -> bool:
+        return codec.startswith("utf-16") or codec.startswith("utf-32")
+
+    @staticmethod
+    def _is_unicode(codec: str) -> bool:
+        return codec.startswith("utf")
+
+    @classmethod
+    def _find_line_matches(cls, line: bytes, codec: str) -> Dict[int, _BadChar]:
+        """Find all matches of BAD_CHARS within line
+
+        Args:
+            line: the input
+            codec: that will be used to convert line/or search string into
+
+        Return:
+            A dictionary with the column offset and the BadASCIIChar
+        """
+        # We try to decode in Unicode to get the correct column offset
+        # if we would use bytes, it could be off because UTF-8 has no fixed length
+        try:
+            line_search = line.decode(codec, errors="strict")
+            search_dict = BAD_ASCII_SEARCH_DICT
+            # If everything is correctly decoded, the BOM will be removed automatically
+            return _map_positions_to_result(line_search, search_dict, "\n")
+        except UnicodeDecodeError:
+            # If we can't decode properly, we simply use bytes, even so the column offsets
+            # might be wrong a bit, but it is still better then nothing
+            line_search_byte = line
+            search_dict_byte = {}
+            for char in BAD_CHARS:
+                # Some characters might not exist in all encodings
+                with contextlib.suppress(UnicodeDecodeError):
+                    search_dict_byte[
+                        _cached_encode_search(char.unescaped, codec)
+                    ] = char
+
+            return _map_positions_to_result(
+                line_search_byte,
+                search_dict_byte,
+                _cached_encode_search("\n", codec),
+                byte_str_length=_byte_to_str_length(codec),
+            )
+
+    def _process_stream(self, stream: io.IOBase) -> None:
+        """Process the stream and add messages if found"""
+
+        # Determine codec
```

Comment:
Each leading comments is a function that could be created (``determine_code``, ``check_codec_validity``, ``check_content`` and ``check_bad_character``)  :)

Reply:
Done